File: agenkit-cpp/examples/infrastructure/batching_example.cpp
/**
 * @file batching_example.cpp
 * @brief Demonstrates batching middleware usage and patterns
 *
 * This example shows:
 * 1. Basic batching setup and configuration
 * 2. Batch size trigger (max_batch_size)
 * 3. Wait time trigger (max_wait_time)
 * 4. Queue size limiting and backpressure
 * 5. Metrics tracking and analysis
 * 6. Concurrent request handling
 * 7. Performance comparison (batched vs unbatched)
 */

#include <agenkit/middleware/batching.hpp>
#include <agenkit/core/agent.hpp>
#include <agenkit/core/message.hpp>
#include <iostream>
#include <iomanip>
#include <chrono>
#include <thread>
#include <vector>
#include <atomic>

using namespace agenkit;
using namespace agenkit::middleware;
using namespace agenkit::core;

void print_separator() {
    std::cout << "\n" << std::string(70, '=') << "\n\n";
}

void print_metrics(const BatchingMetrics& metrics) {
    std::cout << "  Metrics:\n";
    std::cout << "    Total requests:      " << metrics.total_requests.load() << "\n";
    std::cout << "    Total batches:       " << metrics.total_batches.load() << "\n";
    std::cout << "    Successful batches:  " << metrics.successful_batches.load() << "\n";
    std::cout << "    Failed batches:      " << metrics.failed_batches.load() << "\n";
    std::cout << "    Partial batches:     " << metrics.partial_batches.load() << "\n";
    std::cout << "    Min batch size:      " << metrics.min_batch_size.load() << "\n";
    std::cout << "    Max batch size:      " << metrics.max_batch_size.load() << "\n";
    std::cout << "    Avg batch size:      " << std::fixed << std::setprecision(2)
              << metrics.avg_batch_size() << "\n";
    std::cout << "    Avg wait time:       " << std::fixed << std::setprecision(2)
              << metrics.avg_wait_time_ms() << " ms\n";
    std::cout << "    Throughput boost:    " << std::fixed << std::setprecision(2)
              << metrics.throughput_improvement() << "x\n";
}

/// Simulated LLM agent with processing delay
class LLMAgent : public Agent {
public:
    LLMAgent(std::chrono::milliseconds processing_time = std::chrono::milliseconds(100))
        : processing_time_(processing_time) {}

    std::string name() const override {
        return "llm_agent";
    }

    std::future<Result<Message, AgentError>> process(Message message) override {
        request_count_.fetch_add(1);

        // Simulate LLM processing time
        std::this_thread::sleep_for(processing_time_);

        auto response = Message::with_text(
            "assistant",
            "Response to: " + message.content_as_str()
        );

        return make_ready_future(Result<Message, AgentError>::ok(response));
    }

    int request_count() const { return request_count_.load(); }
    void reset_count() { request_count_.store(0); }

private:
    std::chrono::milliseconds processing_time_;
    std::atomic<int> request_count_{0};
};

void example_batch_size_trigger() {
    std::cout << "=== Batch Size Trigger Example ===\n\n";

    auto agent = std::make_shared<LLMAgent>(std::chrono::milliseconds(50));

    auto config = BatchingConfig::builder()
        .max_batch_size(5)
        .max_wait_time(std::chrono::seconds(10))  // Long timeout
        .max_queue_size(100)
        .build();

    auto batching = std::make_shared<BatchingMiddleware>(agent, config);

    std::cout << "Configuration:\n";
    std::cout << "  max_batch_size:  " << config.max_batch_size << "\n";
    std::cout << "  max_wait_time:   " << config.max_wait_time.count() << " ms\n";
    std::cout << "  max_queue_size:  " << config.max_queue_size << "\n\n";

    std::cout << "Sending exactly 5 requests (batch size = 5)...\n";
    auto start = std::chrono::steady_clock::now();

    std::vector<std::future<Result<Message, AgentError>>> futures;
    for (int i = 0; i < 5; i++) {
        auto msg = Message::with_text("user", "Request " + std::to_string(i));
        futures.push_back(batching->process(std::move(msg)));
    }

    // Wait for all responses
    int successes = 0;
    for (auto& fut : futures) {
        auto result = fut.get();
        if (result.is_ok()) {
            successes++;
        }
    }

    auto elapsed = std::chrono::steady_clock::now() - start;
    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();

    std::cout << "\n  Result:\n";
    std::cout << "    Successes:     " << successes << "/" << futures.size() << "\n";
    std::cout << "    Time:          " << elapsed_ms << " ms\n";
    std::cout << "    Agent calls:   " << agent->request_count() << "\n\n";

    // Wait for batch processing to complete
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    print_metrics(batching->metrics());
}

void example_wait_time_trigger() {
    std::cout << "\n=== Wait Time Trigger Example ===\n\n";

    auto agent = std::make_shared<LLMAgent>(std::chrono::milliseconds(50));

    auto config = BatchingConfig::builder()
        .max_batch_size(10)
        .max_wait_time(std::chrono::milliseconds(100))
        .max_queue_size(100)
        .build();

    auto batching = std::make_shared<BatchingMiddleware>(agent, config);

    std::cout << "Configuration:\n";
    std::cout << "  max_batch_size:  " << config.max_batch_size << "\n";
    std::cout << "  max_wait_time:   " << config.max_wait_time.count() << " ms\n\n";

    std::cout << "Sending 3 requests (less than batch size)...\n";
    std::cout << "Batch will trigger after wait time expires.\n";

    auto start = std::chrono::steady_clock::now();

    std::vector<std::future<Result<Message, AgentError>>> futures;
    for (int i = 0; i < 3; i++) {
        auto msg = Message::with_text("user", "Request " + std::to_string(i));
        futures.push_back(batching->process(std::move(msg)));
    }

    // Wait for all responses
    int successes = 0;
    for (auto& fut : futures) {
        auto result = fut.get();
        if (result.is_ok()) {
            successes++;
        }
    }

    auto elapsed = std::chrono::steady_clock::now() - start;
    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();

    std::cout << "\n  Result:\n";
    std::cout << "    Successes:     " << successes << "/" << futures.size() << "\n";
    std::cout << "    Time:          " << elapsed_ms << " ms\n";
    std::cout << "    Agent calls:   " << agent->request_count() << "\n";
    std::cout << "\n  Note: Requests waited ~" << config.max_wait_time.count()
              << " ms for batch timer\n\n";

    std::this_thread::sleep_for(std::chrono::milliseconds(200));
    print_metrics(batching->metrics());
}

void example_concurrent_requests() {
    std::cout << "\n=== Concurrent Requests Example ===\n\n";

    auto agent = std::make_shared<LLMAgent>(std::chrono::milliseconds(30));

    auto config = BatchingConfig::builder()
        .max_batch_size(10)
        .max_wait_time(std::chrono::milliseconds(50))
        .max_queue_size(100)
        .build();

    auto batching = std::make_shared<BatchingMiddleware>(agent, config);

    std::cout << "Launching 50 concurrent requests...\n";
    std::cout << "Batch size: " << config.max_batch_size << "\n\n";

    auto start = std::chrono::steady_clock::now();

    std::vector<std::thread> threads;
    std::atomic<int> success_count{0};

    for (int i = 0; i < 50; i++) {
        threads.emplace_back([&batching, &success_count, i]() {
            auto msg = Message::with_text("user", "Request " + std::to_string(i));
            auto result = batching->process(std::move(msg)).get();
            if (result.is_ok()) {
                success_count.fetch_add(1);
            }
        });
    }

    for (auto& thread : threads) {
        thread.join();
    }

    auto elapsed = std::chrono::steady_clock::now() - start;
    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();

    std::cout << "  Result:\n";
    std::cout << "    Successes:     " << success_count.load() << "/50\n";
    std::cout << "    Time:          " << elapsed_ms << " ms\n";
    std::cout << "    Agent calls:   " << agent->request_count() << "\n\n";

    std::this_thread::sleep_for(std::chrono::milliseconds(200));
    print_metrics(batching->metrics());
}

void example_backpressure() {
    std::cout << "\n=== Backpressure Example ===\n\n";

    auto agent = std::make_shared<LLMAgent>(std::chrono::milliseconds(100));

    auto config = BatchingConfig::builder()
        .max_batch_size(5)
        .max_wait_time(std::chrono::milliseconds(50))
        .max_queue_size(10)  // Small queue
        .build();

    auto batching = std::make_shared<BatchingMiddleware>(agent, config);

    std::cout << "Configuration:\n";
    std::cout << "  max_batch_size:  " << config.max_batch_size << "\n";
    std::cout << "  max_queue_size:  " << config.max_queue_size << " (small)\n\n";

    std::cout << "Rapidly sending 30 requests to small queue...\n";

    std::atomic<int> accepted{0};
    std::atomic<int> rejected{0};

    std::vector<std::future<Result<Message, AgentError>>> futures;
    for (int i = 0; i < 30; i++) {
        auto msg = Message::with_text("user", "Request " + std::to_string(i));
        futures.push_back(batching->process(std::move(msg)));
    }

    for (auto& fut : futures) {
        auto result = fut.get();
        if (result.is_ok()) {
            accepted.fetch_add(1);
        } else {
            rejected.fetch_add(1);
        }
    }

    std::cout << "\n  Result:\n";
    std::cout << "    Accepted:      " << accepted.load() << "\n";
    std::cout << "    Rejected:      " << rejected.load() << " (queue full)\n";
    std::cout << "    Agent calls:   " << agent->request_count() << "\n\n";

    std::cout << "  Note: Rejected requests received immediate error response\n";
    std::cout << "        without waiting or consuming resources.\n\n";

    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    print_metrics(batching->metrics());
}

void example_performance_comparison() {
    std::cout << "\n=== Performance Comparison ===\n\n";

    auto processing_time = std::chrono::milliseconds(50);
    int num_requests = 20;

    // Unbatched version
    std::cout << "Testing unbatched agent...\n";
    auto unbatched = std::make_shared<LLMAgent>(processing_time);

    auto start = std::chrono::steady_clock::now();

    for (int i = 0; i < num_requests; i++) {
        auto msg = Message::with_text("user", "Request " + std::to_string(i));
        auto result = unbatched->process(std::move(msg)).get();
    }

    auto unbatched_time = std::chrono::steady_clock::now() - start;
    auto unbatched_ms = std::chrono::duration_cast<std::chrono::milliseconds>(unbatched_time).count();

    std::cout << "  Time: " << unbatched_ms << " ms\n\n";

    // Batched version
    std::cout << "Testing batched agent...\n";
    auto batched_inner = std::make_shared<LLMAgent>(processing_time);

    auto config = BatchingConfig::builder()
        .max_batch_size(10)
        .max_wait_time(std::chrono::milliseconds(50))
        .build();

    auto batched = std::make_shared<BatchingMiddleware>(batched_inner, config);

    start = std::chrono::steady_clock::now();

    std::vector<std::future<Result<Message, AgentError>>> futures;
    for (int i = 0; i < num_requests; i++) {
        auto msg = Message::with_text("user", "Request " + std::to_string(i));
        futures.push_back(batched->process(std::move(msg)));
    }

    for (auto& fut : futures) {
        fut.get();
    }

    auto batched_time = std::chrono::steady_clock::now() - start;
    auto batched_ms = std::chrono::duration_cast<std::chrono::milliseconds>(batched_time).count();

    std::cout << "  Time: " << batched_ms << " ms\n\n";

    // Wait for batch processing
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    // Results
    std::cout << "  Comparison:\n";
    std::cout << "    Unbatched:     " << unbatched_ms << " ms\n";
    std::cout << "    Batched:       " << batched_ms << " ms\n";
    std::cout << "    Speedup:       " << std::fixed << std::setprecision(2)
              << (double)unbatched_ms / batched_ms << "x\n\n";

    print_metrics(batched->metrics());

    std::cout << "\n  Note: Batching enables parallel processing, reducing overall\n";
    std::cout << "        latency when the agent can handle concurrent requests.\n";
}

int main() {
    std::cout << "Batching Middleware Examples\n";
    std::cout << "============================\n\n";

    try {
        example_batch_size_trigger();
        print_separator();

        example_wait_time_trigger();
        print_separator();

        example_concurrent_requests();
        print_separator();

        example_backpressure();
        print_separator();

        example_performance_comparison();
        print_separator();

        std::cout << "All examples completed successfully!\n\n";
        std::cout << "Key Takeaways:\n";
        std::cout << "1. Batching reduces overhead by processing multiple requests together\n";
        std::cout << "2. max_batch_size controls maximum batch size\n";
        std::cout << "3. max_wait_time ensures latency bounds for small batches\n";
        std::cout << "4. max_queue_size provides backpressure protection\n";
        std::cout << "5. Comprehensive metrics enable monitoring and tuning\n";
        std::cout << "6. Parallel processing within batches improves throughput\n\n";

        return 0;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: agenkit-zig/examples/middleware/batching_example.zig
/// Example demonstrating BatchingMiddleware usage
///
/// Shows how to:
/// 1. Configure batching behavior
/// 2. Process concurrent requests as batches
/// 3. Track batching metrics
/// 4. Observe batch size and wait time

const std = @import("std");
const agktime = @import("../../src/time_compat.zig");
const agenkit = @import("agenkit");

// Slow agent that simulates processing time
const SlowAgent = struct {
    allocator: std.mem.Allocator,
    call_count: std.atomic.Value(u32),
    delay_ms: u64,

    pub fn init(allocator: std.mem.Allocator, delay_ms: u64) !*SlowAgent {
        const self = try allocator.create(SlowAgent);
        self.* = SlowAgent{
            .allocator = allocator,
            .call_count = std.atomic.Value(u32).init(0),
            .delay_ms = delay_ms,
        };
        return self;
    }

    pub fn agent(self: *SlowAgent) agenkit.Agent {
        return agenkit.Agent{
            .ptr = self,
            .vtable = &.{
                .name = nameImpl,
                .capabilities = capabilitiesImpl,
                .process = processImpl,
                .process_stream = processStreamImpl,
                .introspect = introspectImpl,
                .deinit = deinitImpl,
            },
        };
    }

    fn nameImpl(ptr: *anyopaque) []const u8 {
        _ = ptr;
        return "slow_agent";
    }

    fn capabilitiesImpl(ptr: *anyopaque, allocator: std.mem.Allocator) std.mem.Allocator.Error![]const []const u8 {
        _ = ptr;
        const caps = try allocator.alloc([]const u8, 1);
        caps[0] = "slow";
        return caps;
    }

    fn processImpl(ptr: *anyopaque, message: agenkit.Message) agenkit.AgentError!agenkit.Result {
        const self: *SlowAgent = @ptrCast(@alignCast(ptr));

        const count = self.call_count.fetchAdd(1, .monotonic);
        std.debug.print("  [Slow Agent] Processing request #{d}\\n", .{count + 1});

        // Simulate slow processing
        agktime.sleep(self.delay_ms * std.time.ns_per_ms);

        return agenkit.Result{ .ok = message };
    }

    fn introspectImpl(ptr: *anyopaque, allocator: std.mem.Allocator) std.mem.Allocator.Error!agenkit.IntrospectionResult {
        _ = ptr;
        return agenkit.createDefaultIntrospectionResult(allocator, "slow_agent", &.{"slow"});
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *SlowAgent = @ptrCast(@alignCast(ptr));
        self.allocator.destroy(self);
    }
};

// Thread function for concurrent requests
fn sendRequest(args: struct {
    agent: agenkit.Agent,
    allocator: std.mem.Allocator,
    id: u32,
}) void {
    const text = std.fmt.allocPrint(args.allocator, "Request {d}", .{args.id}) catch return;
    defer args.allocator.free(text);

    var message = agenkit.Message.withText(args.allocator, .user, text) catch return;
    defer message.deinit();

    const result = args.agent.process(message) catch |err| {
        std.debug.print("  Thread {d}: ERROR - {}\n", .{ args.id, err });
        return;
    };
    var response = result.unwrap() catch return;
    response.deinit();

    std.debug.print("  Thread {d}: SUCCESS\n", .{args.id});
}


fn processStreamImpl(ptr: *anyopaque, message: Message, callbacks: StreamCallbacks) AgentError!void {
    _ = ptr;
    _ = message;
    callbacks.onError(AgentError.NotImplemented);
}
pub fn main() !void {
    var gpa = std.heap.DebugAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("\n=== Batching Middleware Example ===\n\n", .{});

    // Example 1: Basic batching
    std.debug.print("Example 1: Basic Batching (batch size = 5)\n", .{});
    try basicBatchingExample(allocator);

    std.debug.print("\n", .{});

    // Example 2: Wait time batching
    std.debug.print("Example 2: Wait Time Batching (100ms timeout)\n", .{});
    try waitTimeBatchingExample(allocator);

    std.debug.print("\n", .{});

    // Example 3: Metrics tracking
    std.debug.print("Example 3: Metrics Tracking\n", .{});
    try metricsExample(allocator);

    std.debug.print("\n=== All examples completed successfully ===\n", .{});
}

fn basicBatchingExample(allocator: std.mem.Allocator) !void {
    // Create slow agent (50ms per request)
    var slow = try SlowAgent.init(allocator, 50);
    defer slow.agent().deinit();

    // Configure batching
    const config = agenkit.middleware.BatchingConfig{
        .max_batch_size = 5,
        .max_wait_time_ms = 1000,
        .max_queue_size = 100,
    };
    var batching_agent = try agenkit.middleware.BatchingDecorator.init(allocator, slow.agent(), config);
    defer batching_agent.agent().deinit();

    // Start batch processor
    try batching_agent.start();
    defer batching_agent.stop();

    std.debug.print("  Sending 10 requests concurrently...\n", .{});
    const start_time = agktime.milliTimestamp();

    // Spawn 10 concurrent threads
    var threads: [10]std.Thread = undefined;
    var i: u32 = 0;
    while (i < 10) : (i += 1) {
        threads[i] = try std.Thread.spawn(.{}, sendRequest, .{.{
            .agent = batching_agent.agent(),
            .allocator = allocator,
            .id = i + 1,
        }});
    }

    // Wait for all threads
    for (threads) |thread| {
        thread.join();
    }

    const elapsed_ms = agktime.milliTimestamp() - start_time;
    std.debug.print("\n  Completed in {d}ms\n", .{elapsed_ms});

    const metrics = batching_agent.metrics();
    std.debug.print("  Metrics:\n", .{});
    std.debug.print("    Total requests:      {d}\n", .{metrics.total_requests});
    std.debug.print("    Total batches:       {d}\n", .{metrics.total_batches});
    std.debug.print("    Avg batch size:      {d:.1}\n", .{metrics.avgBatchSize().?});
    std.debug.print("    Agent called only {} times (vs 10 requests)\n", .{metrics.total_batches});
}

fn waitTimeBatchingExample(allocator: std.mem.Allocator) !void {
    // Create slow agent (20ms per request)
    var slow = try SlowAgent.init(allocator, 20);
    defer slow.agent().deinit();

    // Configure batching with wait time
    const config = agenkit.middleware.BatchingConfig{
        .max_batch_size = 10,
        .max_wait_time_ms = 100, // Wait up to 100ms for batch
        .max_queue_size = 100,
    };
    var batching_agent = try agenkit.middleware.BatchingDecorator.init(allocator, slow.agent(), config);
    defer batching_agent.agent().deinit();

    // Start batch processor
    try batching_agent.start();
    defer batching_agent.stop();

    std.debug.print("  Sending 3 requests with delays...\n", .{});
    const start_time = agktime.milliTimestamp();

    // Send 3 requests with 50ms delays between them
    var i: u32 = 0;
    while (i < 3) : (i += 1) {
        const thread = try std.Thread.spawn(.{}, sendRequest, .{.{
            .agent = batching_agent.agent(),
            .allocator = allocator,
            .id = i + 1,
        }});
        thread.detach();

        if (i < 2) {
            agktime.sleep(50 * std.time.ns_per_ms);
        }
    }

    // Wait for processing
    agktime.sleep(300 * std.time.ns_per_ms);

    const elapsed_ms = agktime.milliTimestamp() - start_time;
    std.debug.print("\n  Completed in {d}ms\n", .{elapsed_ms});

    const metrics = batching_agent.metrics();
    std.debug.print("  Metrics:\n", .{});
    std.debug.print("    Total requests:      {d}\n", .{metrics.total_requests});
    std.debug.print("    Total batches:       {d}\n", .{metrics.total_batches});
    std.debug.print("    Avg wait time:       {d:.1}ms\n", .{metrics.avgWaitTime().?});
    std.debug.print("    Batch triggered by timeout (100ms wait)\n", .{});
}

fn metricsExample(allocator: std.mem.Allocator) !void {
    // Create slow agent (30ms per request)
    var slow = try SlowAgent.init(allocator, 30);
    defer slow.agent().deinit();

    // Configure batching
    const config = agenkit.middleware.BatchingConfig{
        .max_batch_size = 5,
        .max_wait_time_ms = 200,
        .max_queue_size = 100,
    };
    var batching_agent = try agenkit.middleware.BatchingDecorator.init(allocator, slow.agent(), config);
    defer batching_agent.agent().deinit();

    // Start batch processor
    try batching_agent.start();
    defer batching_agent.stop();

    std.debug.print("  Sending 25 requests in 5 waves...\n", .{});
    const start_time = agktime.milliTimestamp();

    // Send 5 waves of 5 requests
    var wave: u32 = 0;
    while (wave < 5) : (wave += 1) {
        var i: u32 = 0;
        while (i < 5) : (i += 1) {
            const id = wave * 5 + i + 1;
            const thread = try std.Thread.spawn(.{}, sendRequest, .{.{
                .agent = batching_agent.agent(),
                .allocator = allocator,
                .id = id,
            }});
            thread.detach();
        }

        // Wait between waves
        if (wave < 4) {
            agktime.sleep(100 * std.time.ns_per_ms);
        }
    }

    // Wait for all processing
    agktime.sleep(500 * std.time.ns_per_ms);

    const elapsed_ms = agktime.milliTimestamp() - start_time;
    const metrics = batching_agent.metrics();

    std.debug.print("\n  Final Metrics:\n", .{});
    std.debug.print("    Total requests:        {d}\n", .{metrics.total_requests});
    std.debug.print("    Total batches:         {d}\n", .{metrics.total_batches});
    std.debug.print("    Successful batches:    {d}\n", .{metrics.successful_batches});
    std.debug.print("    Failed batches:        {d}\n", .{metrics.failed_batches});
    std.debug.print("    Partial batches:       {d}\n", .{metrics.partial_batches});
    std.debug.print("    Min batch size:        {d}\n", .{metrics.min_batch_size.?});
    std.debug.print("    Max batch size:        {d}\n", .{metrics.max_batch_size.?});
    std.debug.print("    Avg batch size:        {d:.2}\n", .{metrics.avgBatchSize().?});
    std.debug.print("    Avg wait time:         {d:.1}ms\n", .{metrics.avgWaitTime().?});
    std.debug.print("    Total time:            {d}ms\n", .{elapsed_ms});
    std.debug.print("\n    Throughput improvement: {d:.1}x\n", .{
        @as(f64, @floatFromInt(metrics.total_requests)) / @as(f64, @floatFromInt(metrics.total_batches)),
    });
}
